shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmPluginDataHandlerTest.java

```diff
@@ -0,0 +1,203 @@
+
+    @BeforeEach
+    public void setUp() {
+        this.ruleData = mock(RuleData.class);
+        this.pluginData = mock(PluginData.class);
+        this.selectorData = mock(SelectorData.class);
+        this.testWasmPluginDataHandler = new TestWasmPluginDataHandler();
+        this.pluginDataHandler = () -> "SHENYU";
+        when(ruleData.getId()).thenReturn("SHENYU");
+        when(pluginData.getId()).thenReturn("SHENYU");
+        when(selectorData.getId()).thenReturn("SHENYU");
+    }
+
+    /**
+     * The handler plugin test.
+     */
+    @Test
+    public void handlerPluginTest() {
+        pluginDataHandler.handlerPlugin(pluginData);
+        testWasmPluginDataHandler.handlerPlugin(pluginData);
+        assertEquals(testWasmPluginDataHandler.getPluginDataMap().get(pluginData.getId()).getId(), pluginData.getId());
+    }
+
+    /**
+     * The remove plugin test.
+     */
+    @Test
+    public void removePluginTest() {
+        pluginDataHandler.removePlugin(pluginData);
+        testWasmPluginDataHandler.handlerPlugin(pluginData);
+        testWasmPluginDataHandler.removePlugin(pluginData);
+        assertNull(testWasmPluginDataHandler.getPluginDataMap().get(pluginData.getId()));
+    }
+
+    /**
+     * The handler selector test.
+     */
+    @Test
+    public void handlerSelectorTest() {
+        pluginDataHandler.handlerSelector(selectorData);
+        testWasmPluginDataHandler.handlerSelector(selectorData);
+        assertEquals(testWasmPluginDataHandler.getSelectorDataMap().get(selectorData.getId()).getId(), selectorData.getId());
+    }
+
+    /**
+     * The remove selector test.
+     */
+    @Test
+    public void removeSelectorTest() {
+        pluginDataHandler.removeSelector(selectorData);
+        testWasmPluginDataHandler.handlerSelector(selectorData);
+        testWasmPluginDataHandler.removeSelector(selectorData);
+        assertNull(testWasmPluginDataHandler.getSelectorDataMap().get(selectorData.getId()));
+    }
+
+    /**
+     * The handler rule test.
+     */
+    @Test
+    public void handlerRuleTest() {
+        pluginDataHandler.handlerRule(ruleData);
+        testWasmPluginDataHandler.handlerRule(ruleData);
+        assertEquals(testWasmPluginDataHandler.getRuleDataMap().get(ruleData.getId()).getId(), ruleData.getId());
+    }
+
+    /**
+     * The remove rule test.
+     */
+    @Test
+    public void removeRuleTest() {
+        pluginDataHandler.removeRule(ruleData);
+        testWasmPluginDataHandler.handlerRule(ruleData);
+        testWasmPluginDataHandler.removeRule(ruleData);
+        assertNull(testWasmPluginDataHandler.getRuleDataMap().get(ruleData.getId()));
+    }
+
+    /**
+     * The plugin named test.
+     */
+    @Test
+    public void pluginNamedTest() {
+        assertEquals("SHENYU", pluginDataHandler.pluginNamed());
+        assertEquals("SHENYU_TEST", testWasmPluginDataHandler.pluginNamed());
+    }
+
+    static class TestWasmPluginDataHandler implements PluginDataHandler {
```

Comment:
Need to inherit `AbstractWasmPluginDataHandler`, now this implementation will not load&call the WASM function.

Reply:
Could you please take a look and tell why is this error coming @loongs-zhang .The wasm file is located in the resources folder so this error shouldn't be coming?
```Powershell
"C:\Program Files\Java\jdk-21\bin\java.exe" -ea -Didea.test.cyclic.buffer.size=1048576 "-javaagent:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2023.3.2\lib\idea_rt.jar=55442:C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2023.3.2\bin" -Dfile.encoding=UTF-8 -Dsun.stdout.encoding=UTF-8 -Dsun.stderr.encoding=UTF-8 -classpath "C:\Users\Divyansh\.m2\repository\org\junit\platform\junit-platform-launcher\1.8.2\junit-platform-launcher-1.8.2.jar;C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2023.3.2\lib\idea_rt.jar;C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2023.3.2\plugins\junit\lib\junit5-rt.jar;C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2023.3.2\plugins\junit\lib\junit-rt.jar;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-plugin\shenyu-plugin-wasm-base\target\test-classes;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-plugin\shenyu-plugin-wasm-base\target\classes;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-plugin\shenyu-plugin-wasm-api\target\classes;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-plugin\shenyu-plugin-api\target\classes;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-sync-data-center\shenyu-sync-data-api\target\classes;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-common\target\classes;C:\Users\Divyansh\.m2\repository\com\google\code\gson\gson\2.9.0\gson-2.9.0.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\datatype\jackson-datatype-jdk8\2.13.5\jackson-datatype-jdk8-2.13.5.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\datatype\jackson-datatype-jsr310\2.13.5\jackson-datatype-jsr310-2.13.5.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\module\jackson-module-parameter-names\2.13.5\jackson-module-parameter-names-2.13.5.jar;C:\Users\Divyansh\.m2\repository\commons-codec\commons-codec\1.15\commons-codec-1.15.jar;C:\Users\Divyansh\.m2\repository\org\ow2\asm\asm-tree\9.2\asm-tree-9.2.jar;C:\Users\Divyansh\.m2\repository\org\ow2\asm\asm\9.2\asm-9.2.jar;C:\Users\Divyansh\.m2\repository\io\projectreactor\netty\reactor-netty-core\1.0.38\reactor-netty-core-1.0.38.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-handler\4.1.100.Final\netty-handler-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-common\4.1.100.Final\netty-common-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-resolver\4.1.100.Final\netty-resolver-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-buffer\4.1.100.Final\netty-buffer-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-transport\4.1.100.Final\netty-transport-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-transport-native-unix-common\4.1.100.Final\netty-transport-native-unix-common-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-codec\4.1.100.Final\netty-codec-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-handler-proxy\4.1.100.Final\netty-handler-proxy-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-codec-socks\4.1.100.Final\netty-codec-socks-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-resolver-dns\4.1.100.Final\netty-resolver-dns-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-codec-dns\4.1.100.Final\netty-codec-dns-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-resolver-dns-native-macos\4.1.100.Final\netty-resolver-dns-native-macos-4.1.100.Final-osx-x86_64.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-resolver-dns-classes-macos\4.1.100.Final\netty-resolver-dns-classes-macos-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-transport-native-epoll\4.1.100.Final\netty-transport-native-epoll-4.1.100.Final-linux-x86_64.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-transport-classes-epoll\4.1.100.Final\netty-transport-classes-epoll-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\projectreactor\netty\reactor-netty-http\1.0.38\reactor-netty-http-1.0.38.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-codec-http\4.1.100.Final\netty-codec-http-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\io\netty\netty-codec-http2\4.1.100.Final\netty-codec-http2-4.1.100.Final.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-webflux\5.3.30\spring-webflux-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-beans\5.3.30\spring-beans-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-web\5.3.30\spring-web-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-context\5.3.30\spring-context-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-aop\5.3.30\spring-aop-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-expression\5.3.30\spring-expression-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\apache\commons\commons-lang3\3.12.0\commons-lang3-3.12.0.jar;C:\Users\Divyansh\.m2\repository\io\github\kawamuray\wasmtime\wasmtime-java\0.19.0\wasmtime-java-0.19.0.jar;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-plugin\shenyu-plugin-base\target\classes;C:\Users\Divyansh\.m2\repository\com\github\ben-manes\caffeine\caffeine\2.9.3\caffeine-2.9.3.jar;C:\Users\Divyansh\Documents\GitHub\shenyu\shenyu-spi\target\classes;C:\Users\Divyansh\.m2\repository\org\apache\commons\commons-collections4\4.4\commons-collections4-4.4.jar;C:\Users\Divyansh\.m2\repository\ch\qos\logback\logback-classic\1.2.12\logback-classic-1.2.12.jar;C:\Users\Divyansh\.m2\repository\ch\qos\logback\logback-core\1.2.12\logback-core-1.2.12.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\module\jackson-module-kotlin\2.13.5\jackson-module-kotlin-2.13.5.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.13.5\jackson-databind-2.13.5.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\core\jackson-core\2.13.5\jackson-core-2.13.5.jar;C:\Users\Divyansh\.m2\repository\com\fasterxml\jackson\core\jackson-annotations\2.13.5\jackson-annotations-2.13.5.jar;C:\Users\Divyansh\.m2\repository\org\jetbrains\kotlin\kotlin-reflect\1.6.21\kotlin-reflect-1.6.21.jar;C:\Users\Divyansh\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.21\kotlin-stdlib-1.6.21.jar;C:\Users\Divyansh\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.6.21\kotlin-stdlib-common-1.6.21.jar;C:\Users\Divyansh\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-test\5.3.30\spring-test-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-core\5.3.30\spring-core-5.3.30.jar;C:\Users\Divyansh\.m2\repository\org\springframework\spring-jcl\5.3.30\spring-jcl-5.3.30.jar;C:\Users\Divyansh\.m2\repository\io\projectreactor\reactor-test\3.4.33\reactor-test-3.4.33.jar;C:\Users\Divyansh\.m2\repository\io\projectreactor\reactor-core\3.4.33\reactor-core-3.4.33.jar;C:\Users\Divyansh\.m2\repository\org\reactivestreams\reactive-streams\1.0.4\reactive-streams-1.0.4.jar;C:\Users\Divyansh\.m2\repository\org\assertj\assertj-core\3.22.0\assertj-core-3.22.0.jar;C:\Users\Divyansh\.m2\repository\com\google\guava\guava\32.0.0-jre\guava-32.0.0-jre.jar;C:\Users\Divyansh\.m2\repository\com\google\guava\failureaccess\1.0.1\failureaccess-1.0.1.jar;C:\Users\Divyansh\.m2\repository\com\google\guava\listenablefuture\9999.0-empty-to-avoid-conflict-with-guava\listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar;C:\Users\Divyansh\.m2\repository\com\google\code\findbugs\jsr305\3.0.2\jsr305-3.0.2.jar;C:\Users\Divyansh\.m2\repository\org\checkerframework\checker-qual\3.33.0\checker-qual-3.33.0.jar;C:\Users\Divyansh\.m2\repository\com\google\errorprone\error_prone_annotations\2.18.0\error_prone_annotations-2.18.0.jar;C:\Users\Divyansh\.m2\repository\com\google\j2objc\j2objc-annotations\2.8\j2objc-annotations-2.8.jar;C:\Users\Divyansh\.m2\repository\org\slf4j\slf4j-api\1.7.36\slf4j-api-1.7.36.jar;C:\Users\Divyansh\.m2\repository\org\slf4j\jcl-over-slf4j\1.7.36\jcl-over-slf4j-1.7.36.jar;C:\Users\Divyansh\.m2\repository\org\junit\jupiter\junit-jupiter\5.8.2\junit-jupiter-5.8.2.jar;C:\Users\Divyansh\.m2\repository\org\junit\jupiter\junit-jupiter-api\5.8.2\junit-jupiter-api-5.8.2.jar;C:\Users\Divyansh\.m2\repository\org\opentest4j\opentest4j\1.2.0\opentest4j-1.2.0.jar;C:\Users\Divyansh\.m2\repository\org\junit\platform\junit-platform-commons\1.8.2\junit-platform-commons-1.8.2.jar;C:\Users\Divyansh\.m2\repository\org\junit\jupiter\junit-jupiter-params\5.8.2\junit-jupiter-params-5.8.2.jar;C:\Users\Divyansh\.m2\repository\org\junit\jupiter\junit-jupiter-engine\5.8.2\junit-jupiter-engine-5.8.2.jar;C:\Users\Divyansh\.m2\repository\org\junit\vintage\junit-vintage-engine\5.8.2\junit-vintage-engine-5.8.2.jar;C:\Users\Divyansh\.m2\repository\org\junit\platform\junit-platform-engine\1.8.2\junit-platform-engine-1.8.2.jar;C:\Users\Divyansh\.m2\repository\junit\junit\4.13.2\junit-4.13.2.jar;C:\Users\Divyansh\.m2\repository\org\apiguardian\apiguardian-api\1.1.2\apiguardian-api-1.1.2.jar;C:\Users\Divyansh\.m2\repository\org\hamcrest\hamcrest-library\2.2\hamcrest-library-2.2.jar;C:\Users\Divyansh\.m2\repository\org\hamcrest\hamcrest-core\2.2\hamcrest-core-2.2.jar;C:\Users\Divyansh\.m2\repository\org\mockito\mockito-junit-jupiter\3.5.15\mockito-junit-jupiter-3.5.15.jar;C:\Users\Divyansh\.m2\repository\org\mockito\mockito-core\3.5.15\mockito-core-3.5.15.jar;C:\Users\Divyansh\.m2\repository\net\bytebuddy\byte-buddy\1.14.11\byte-buddy-1.14.11.jar;C:\Users\Divyansh\.m2\repository\net\bytebuddy\byte-buddy-agent\1.12.23\byte-buddy-agent-1.12.23.jar;C:\Users\Divyansh\.m2\repository\org\objenesis\objenesis\3.1\objenesis-3.1.jar;C:\Users\Divyansh\.m2\repository\org\mockito\mockito-inline\3.5.15\mockito-inline-3.5.15.jar;C:\Users\Divyansh\.m2\repository\org\awaitility\awaitility\4.0.3\awaitility-4.0.3.jar;C:\Users\Divyansh\.m2\repository\org\hamcrest\hamcrest\2.2\hamcrest-2.2.jar;C:\Users\Divyansh\.m2\repository\javax\xml\bind\jaxb-api\2.3.1\jaxb-api-2.3.1.jar;C:\Users\Divyansh\.m2\repository\javax\activation\javax.activation-api\1.2.0\javax.activation-api-1.2.0.jar" com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit5 org.apache.shenyu.plugin.wasm.base.handler.AbstractWasmPluginDataHandlerTest
WARNING: A Java agent has been loaded dynamically (C:\Users\Divyansh\.m2\repository\net\bytebuddy\byte-buddy-agent\1.12.23\byte-buddy-agent-1.12.23.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release
Java HotSpot(TM) 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
15:48:43.553 [main] DEBUG io.github.kawamuray.wasmtime.NativeLibraryLoader - Loading Wasmtime JNI library from C:\Users\Divyansh\AppData\Local\Temp\wasmtime_jni_0.19.0_windows_x86_64623547734575093793.dll

org.apache.shenyu.plugin.wasm.api.exception.ShenyuWasmInitException: Can't find wasm file: org.apache.shenyu.plugin.wasm.base.handler.AbstractWasmPluginDataHandlerTest$TestWasmPluginDataHandler.wasm

```

Reply:
@loongs-zhang Could you please take a look at the below error? It says `Can't find the wasm file`. It's coming when I made below  change but since the wasm file is located in resources folder I am sure why this error is coming.The same error is coming in #5452 
```diff
-    static class TestWasmPluginDataHandler implements PluginDataHandler{ 
+    static class TestWasmPluginDataHandler extends AbstractWasmPluginDataHandler {

``` 

```powershell
WARNING: A Java agent has been loaded dynamically (C:\Users\Divyansh\.m2\repository\net\bytebuddy\byte-buddy-agent\1.12.23\byte-buddy-agent-1.12.23.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release
Java HotSpot(TM) 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
17:18:08.304 [main] DEBUG io.github.kawamuray.wasmtime.NativeLibraryLoader - Loading Wasmtime JNI library from C:\Users\Divyansh\AppData\Local\Temp\wasmtime_jni_0.19.0_windows_x86_6411715251905330268411.dll

org.apache.shenyu.plugin.wasm.api.exception.ShenyuWasmInitException: Can't find wasm file: org.apache.shenyu.plugin.wasm.base.handler.AbstractWasmPluginDataHandlerTest$TestWasmPluginDataHandler.wasm
```

Reply:
The class is `shenyu-plugin/shenyu-plugin-wasm-base/src/test/java/org/apache/shenyu/plugin/wasm/base/handler/AbstractWasmPluginDataHandlerTest.java`, so the correct WASM file name is `org.apache.shenyu.plugin.wasm.base.handler.AbstractWasmPluginDataHandlerTest$TestWasmPluginDataHandler.wasm`.

Reply:
> shenyu-plugin/shenyu-plugin-wasm-base/src/test/resources/org.apache.shenyu.plugin.wasm.base.AbstractWasmPluginDataHandlerTest$TestWasmPluginDataHandler.wasm

You lost the `handler `.